File: 54_Spiral Matrix.java
// 时间复杂度：O(mn)，其中 m 和 n 分别为 matrix 的行数和列数。
// 空间复杂度：O(1)。返回值不计入。

class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int[][] dir = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        List<Integer> res = new ArrayList<>();
        int row = matrix.length, col = matrix[0].length, r = 0, c = 0, d = 0;

        for (int i = 0; i < row*col; i++) {
            // 1. record & update visited
            res.add(matrix[r][c]);
            matrix[r][c] = -101;
            // 2. shift to the next valid position
            int nextr = r + dir[d][0], nextc = c + dir[d][1];
            // 3. turn 90D clockwise if out of boundary or visited
            if (nextr < 0 || nextr >= row || nextc < 0 || nextc >= col || matrix[nextr][nextc] == -101) {
                d = (d+1)%4;
            }
            r = r + dir[d][0];
            c = c + dir[d][1];
        }

        return res;
    }
}

File: 54_Spiral Matrix.py
# 螺旋结构确保当你撞到墙壁或访问过的牢房并顺时针转90°时，新路径是开放的

class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        dirs = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        res = []
        row, col = len(matrix), len(matrix[0])
        r = c = d = 0

        for _ in range(row*col):
            # update res
            res.append(matrix[r][c])
            # mark as visited
            matrix[r][c] = -inf
            # explore the next valid position
            if not (0 <= r + dirs[d][0] < row and 0 <= c + dirs[d][1] < col and matrix[r+dirs[d][0]][c+dirs[d][1]] != -inf): d = (d + 1) % 4
            r, c = r + dirs[d][0], c + dirs[d][1]
        return res